File: backend/src/schemas/reporting.py
from datetime import date
from decimal import Decimal
from enum import Enum
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


class ReportPeriodType(str, Enum):
    MONTHLY = "MONTHLY"
    QUARTERLY = "QUARTERLY"
    YEARLY = "YEARLY"
    CUSTOM = "CUSTOM"


class FinancialReportQuery(BaseModel):
    period_type: ReportPeriodType = ReportPeriodType.MONTHLY
    start_date: date
    end_date: date
    compare_with: Optional[str] = None
    project_id: Optional[str] = None


class GeneralLedgerQuery(BaseModel):
    account_code: str
    start_date: date
    end_date: date
    project_id: Optional[str] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=50, ge=1, le=500)


class ReportLineItem(BaseModel):
    account_code: Optional[str] = None
    line_name: str
    amount: Decimal
    comparative_amount: Optional[Decimal] = None
    variance_amount: Optional[Decimal] = None
    variance_percentage: Optional[Decimal] = None
    drill_down_supported: bool = True


class ReportSection(BaseModel):
    section_code: str
    section_name: str
    lines: List[ReportLineItem] = []
    subtotal: Decimal = Decimal("0.00")
    comparative_subtotal: Optional[Decimal] = None


# Profit & Loss
class ProfitLossReportResponse(BaseModel):
    organization_name: str
    period_label: str
    start_date: date
    end_date: date
    generated_at: str
    revenue_section: ReportSection
    cogs_section: ReportSection
    gross_profit: Decimal
    gross_margin_percentage: Decimal
    operating_expenses_section: ReportSection
    operating_profit: Decimal
    other_income_expense_section: ReportSection
    earnings_before_tax: Decimal
    tax_expense: Decimal
    net_profit: Decimal


# Balance Sheet
class BalanceSheetReportResponse(BaseModel):
    organization_name: str
    as_of_date: date
    generated_at: str
    current_assets: ReportSection
    fixed_assets: ReportSection
    total_assets: Decimal
    current_liabilities: ReportSection
    long_term_liabilities: ReportSection
    total_liabilities: Decimal
    equity: ReportSection
    total_equity: Decimal
    total_liabilities_and_equity: Decimal
    is_balanced: bool
    balancing_difference: Decimal
    integrity_status: str  # "VALID" | "INTEGRITY_ERROR"


# Trial Balance
class TrialBalanceLine(BaseModel):
    account_code: str
    account_name: str
    account_type: str
    normal_balance: str
    opening_debit: Decimal
    opening_credit: Decimal
    period_debit: Decimal
    period_credit: Decimal
    ending_debit: Decimal
    ending_credit: Decimal


class TrialBalanceResponse(BaseModel):
    organization_name: str
    as_of_date: date
    start_date: date
    end_date: date
    lines: List[TrialBalanceLine] = []
    total_opening_debit: Decimal = Decimal("0.00")
    total_opening_credit: Decimal = Decimal("0.00")
    total_period_debit: Decimal = Decimal("0.00")
    total_period_credit: Decimal = Decimal("0.00")
    total_ending_debit: Decimal = Decimal("0.00")
    total_ending_credit: Decimal = Decimal("0.00")
    is_balanced: bool = True
    difference: Decimal = Decimal("0.00")


# General Ledger
class GeneralLedgerEntry(BaseModel):
    date: date
    journal_entry_id: str
    journal_entry_number: str
    transaction_id: Optional[str] = None
    description: str
    project_code: Optional[str] = None
    project_name: Optional[str] = None
    debit: Decimal
    credit: Decimal
    running_balance: Decimal
    document_ids: List[str] = []


class GeneralLedgerResponse(BaseModel):
    organization_name: str
    account_code: str
    account_name: str
    account_type: str
    normal_balance: str
    start_date: date
    end_date: date
    opening_balance: Decimal
    total_debit: Decimal
    total_credit: Decimal
    closing_balance: Decimal
    entries: List[GeneralLedgerEntry] = []
    total_records: int
    page: int
    page_size: int


# Integrity Diagnostics
class IntegrityCheckItem(BaseModel):
    check_name: str
    status: str  # "PASS" | "FAIL"
    left_value: Decimal
    right_value: Decimal
    discrepancy: Decimal
    message: str


class IntegrityReportResponse(BaseModel):
    organization_name: str
    as_of_date: date
    overall_status: str  # "VALID" | "INTEGRITY_ERROR"
    checks: List[IntegrityCheckItem] = []


# Cash Flow (Direct Method)
class CashFlowReportResponse(BaseModel):
    organization_name: str
    period_label: str
    start_date: date
    end_date: date
    opening_cash_balance: Decimal
    operating_activities: ReportSection
    net_operating_cash: Decimal
    investing_activities: ReportSection
    net_investing_cash: Decimal
    financing_activities: ReportSection
    net_financing_cash: Decimal
    net_cash_change: Decimal
    closing_cash_balance: Decimal
    unclassified_cash_activities: Optional[ReportSection] = None


# AR Aging
class AgingBucketSummary(BaseModel):
    current: Decimal
    days_1_30: Decimal
    days_31_60: Decimal
    days_61_90: Decimal
    days_over_90: Decimal
    total: Decimal


class ARAgingInvoiceLine(BaseModel):
    customer_id: str
    customer_name: str
    project_code: Optional[str] = None
    project_name: Optional[str] = None
    invoice_number: str
    invoice_date: date
    due_date: date
    days_overdue: int
    total_amount: Decimal
    paid_amount: Decimal
    outstanding_amount: Decimal
    bucket: str  # "CURRENT", "1_30", "31_60", "61_90", "OVER_90"


class ARAgingReportResponse(BaseModel):
    organization_name: str
    as_of_date: date
    summary: AgingBucketSummary
    invoices: List[ARAgingInvoiceLine] = []


# AP Aging
class APAgingBillLine(BaseModel):
    vendor_id: str
    vendor_name: str
    project_code: Optional[str] = None
    project_name: Optional[str] = None
    bill_number: str
    bill_date: date
    due_date: date
    days_overdue: int
    total_amount: Decimal
    paid_amount: Decimal
    outstanding_amount: Decimal
    bucket: str


class APAgingReportResponse(BaseModel):
    organization_name: str
    as_of_date: date
    summary: AgingBucketSummary
    bills: List[APAgingBillLine] = []
    unsettled_advances_total: Decimal = Decimal("0.00")


# Project Profitability
class ProjectCostCategoryLine(BaseModel):
    cost_category: str  # "MAT", "SUB", etc.
    category_name: str
    amount: Decimal


class ProjectProfitabilityReportResponse(BaseModel):
    organization_name: str
    project_id: str
    project_code: str
    project_name: str
    client_name: Optional[str] = None
    status: str
    original_contract_value: Decimal
    variation_orders_value: Decimal
    revised_contract_value: Decimal
    revenue_recognized: Decimal
    cost_breakdown: List[ProjectCostCategoryLine] = []
    total_project_cost: Decimal
    gross_profit: Decimal
    gross_margin_percentage: Decimal


# Project Cash Position
class ProjectCashPositionReportResponse(BaseModel):
    organization_name: str
    project_id: str
    project_code: str
    project_name: str
    invoiced_amount: Decimal
    cash_received: Decimal
    receivable_outstanding: Decimal
    cash_spent: Decimal
    net_cash_position: Decimal
    is_surplus: bool
    notice_message: str = "Laba Proyek (Akrual) Berbeda dengan Posisi Kas Proyek (Likuiditas)."


# Budget vs Actual
class BudgetVsActualLine(BaseModel):
    cost_category: str
    category_name: str
    budget_amount: Decimal
    actual_amount: Decimal
    variance_amount: Decimal
    variance_percentage: Decimal
    status: str  # "NORMAL", "WARNING", "OVERBUDGET"


class BudgetVsActualReportResponse(BaseModel):
    organization_name: str
    project_id: str
    project_code: str
    project_name: str
    has_budget: bool
    budget_status_label: str
    total_budget: Decimal
    total_actual: Decimal
    total_variance: Decimal
    consumption_percentage: Decimal
    lines: List[BudgetVsActualLine] = []


# Executive Management Dashboard
class DashboardSummaryResponse(BaseModel):
    organization_name: str
    as_of_date: date
    cash_and_bank_balance: Decimal
    accounts_receivable_outstanding: Decimal
    accounts_payable_outstanding: Decimal
    revenue_ytd: Decimal
    net_profit_ytd: Decimal
    estimated_monthly_burn_rate: Decimal
    cash_runway_months: Optional[Decimal] = None
    active_projects_count: int
    review_queue_pending_count: int
    integrity_status: str  # "VALID" | "INTEGRITY_ERROR"

File: frontend/src/types/reporting.ts
export interface ReportLineItem {
  account_code?: string;
  line_name: string;
  amount: number | string;
  comparative_amount?: number | string | null;
  variance_amount?: number | string | null;
  variance_percentage?: number | string | null;
  drill_down_supported: boolean;
}

export interface ReportSection {
  section_code: string;
  section_name: string;
  lines: ReportLineItem[];
  subtotal: number | string;
  comparative_subtotal?: number | string | null;
}

export interface ProfitLossReportResponse {
  organization_name: string;
  period_label: string;
  start_date: string;
  end_date: string;
  generated_at: string;
  revenue_section: ReportSection;
  cogs_section: ReportSection;
  gross_profit: number | string;
  gross_margin_percentage: number | string;
  operating_expenses_section: ReportSection;
  operating_profit: number | string;
  other_income_expense_section: ReportSection;
  earnings_before_tax: number | string;
  tax_expense: number | string;
  net_profit: number | string;
}

export interface BalanceSheetReportResponse {
  organization_name: string;
  as_of_date: string;
  generated_at: string;
  current_assets: ReportSection;
  fixed_assets: ReportSection;
  total_assets: number | string;
  current_liabilities: ReportSection;
  long_term_liabilities: ReportSection;
  total_liabilities: number | string;
  equity: ReportSection;
  total_equity: number | string;
  total_liabilities_and_equity: number | string;
  is_balanced: boolean;
  balancing_difference: number | string;
  integrity_status: 'VALID' | 'INTEGRITY_ERROR';
}

export interface TrialBalanceLine {
  account_code: string;
  account_name: string;
  account_type: string;
  normal_balance: string;
  opening_debit: number | string;
  opening_credit: number | string;
  period_debit: number | string;
  period_credit: number | string;
  ending_debit: number | string;
  ending_credit: number | string;
}

export interface TrialBalanceResponse {
  organization_name: string;
  as_of_date: string;
  start_date: string;
  end_date: string;
  lines: TrialBalanceLine[];
  total_opening_debit: number | string;
  total_opening_credit: number | string;
  total_period_debit: number | string;
  total_period_credit: number | string;
  total_ending_debit: number | string;
  total_ending_credit: number | string;
  is_balanced: boolean;
  difference: number | string;
}

export interface GeneralLedgerEntry {
  date: string;
  journal_entry_id: string;
  journal_entry_number: string;
  transaction_id?: string | null;
  description: string;
  project_code?: string | null;
  project_name?: string | null;
  debit: number | string;
  credit: number | string;
  running_balance: number | string;
  document_ids: string[];
}

export interface GeneralLedgerResponse {
  organization_name: string;
  account_code: string;
  account_name: string;
  account_type: string;
  normal_balance: string;
  start_date: string;
  end_date: string;
  opening_balance: number | string;
  total_debit: number | string;
  total_credit: number | string;
  closing_balance: number | string;
  entries: GeneralLedgerEntry[];
  total_records: number;
  page: number;
  page_size: number;
}

export interface IntegrityCheckItem {
  check_name: string;
  status: 'PASS' | 'FAIL';
  left_value: number | string;
  right_value: number | string;
  discrepancy: number | string;
  message: string;
}

export interface IntegrityReportResponse {
  organization_name: string;
  as_of_date: string;
  overall_status: 'VALID' | 'INTEGRITY_ERROR';
  checks: IntegrityCheckItem[];
}

export interface CashFlowReportResponse {
  organization_name: string;
  period_label: string;
  start_date: string;
  end_date: string;
  opening_cash_balance: number | string;
  operating_activities: ReportSection;
  net_operating_cash: number | string;
  investing_activities: ReportSection;
  net_investing_cash: number | string;
  financing_activities: ReportSection;
  net_financing_cash: number | string;
  net_cash_change: number | string;
  closing_cash_balance: number | string;
  unclassified_cash_activities?: ReportSection | null;
}

export interface AgingBucketSummary {
  current: number | string;
  days_1_30: number | string;
  days_31_60: number | string;
  days_61_90: number | string;
  days_over_90: number | string;
  total: number | string;
}

export interface ARAgingInvoiceLine {
  customer_id: string;
  customer_name: string;
  project_code?: string | null;
  project_name?: string | null;
  invoice_number: string;
  invoice_date: string;
  due_date: string;
  days_overdue: number;
  total_amount: number | string;
  paid_amount: number | string;
  outstanding_amount: number | string;
  bucket: string;
}

export interface ARAgingReportResponse {
  organization_name: string;
  as_of_date: string;
  summary: AgingBucketSummary;
  invoices: ARAgingInvoiceLine[];
}

export interface APAgingBillLine {
  vendor_id: string;
  vendor_name: string;
  project_code?: string | null;
  project_name?: string | null;
  bill_number: string;
  bill_date: string;
  due_date: string;
  days_overdue: number;
  total_amount: number | string;
  paid_amount: number | string;
  outstanding_amount: number | string;
  bucket: string;
}

export interface APAgingReportResponse {
  organization_name: string;
  as_of_date: string;
  summary: AgingBucketSummary;
  bills: APAgingBillLine[];
  unsettled_advances_total: number | string;
}

export interface ProjectCostCategoryLine {
  cost_category: string;
  category_name: string;
  amount: number | string;
}

export interface ProjectProfitabilityReportResponse {
  organization_name: string;
  project_id: string;
  project_code: string;
  project_name: string;
  client_name?: string | null;
  status: string;
  original_contract_value: number | string;
  variation_orders_value: number | string;
  revised_contract_value: number | string;
  revenue_recognized: number | string;
  cost_breakdown: ProjectCostCategoryLine[];
  total_project_cost: number | string;
  gross_profit: number | string;
  gross_margin_percentage: number | string;
}

export interface ProjectCashPositionReportResponse {
  organization_name: string;
  project_id: string;
  project_code: string;
  project_name: string;
  invoiced_amount: number | string;
  cash_received: number | string;
  receivable_outstanding: number | string;
  cash_spent: number | string;
  net_cash_position: number | string;
  is_surplus: boolean;
  notice_message: string;
}

export interface BudgetVsActualLine {
  cost_category: string;
  category_name: string;
  budget_amount: number | string;
  actual_amount: number | string;
  variance_amount: number | string;
  variance_percentage: number | string;
  status: string;
}

export interface BudgetVsActualReportResponse {
  organization_name: string;
  project_id: string;
  project_code: string;
  project_name: string;
  has_budget: boolean;
  budget_status_label: string;
  total_budget: number | string;
  total_actual: number | string;
  total_variance: number | string;
  consumption_percentage: number | string;
  lines: BudgetVsActualLine[];
}

export interface DashboardSummaryResponse {
  organization_name: string;
  as_of_date: string;
  cash_and_bank_balance: number | string;
  cash_in_period: number | string;
  cash_out_period: number | string;
  net_cash_flow: number | string;
  unallocated_cash: number | string;
  project_spending: number | string;
  accounts_receivable_outstanding: number | string;
  accounts_payable_outstanding: number | string;
  revenue_ytd: number | string;
  net_profit_ytd: number | string;
  estimated_monthly_burn_rate: number | string;
  cash_runway_months?: number | string | null;
  active_projects_count: number;
  review_queue_pending_count: number;
  integrity_status: 'VALID' | 'INTEGRITY_ERROR';
}


